core/src/autogluon/core/models/abstract/abstract_model.py

```diff
@@ -539,61 +539,73 @@ def fit_with_prune(self, X, y, X_val, y_val, max_num_fit=3, prune_threshold=0.,
+        3. If self.weight_evaluation==True pass sample_weight param for scoring
 
         FIXME
         1. Does not work with S3 paths (no problem locally)
         """
-        if X_val is None or y_val is None:
-            raise AssertionError(f'fit_with_prune must be called with validation data.')
+        logger.log(20, f"Using Iterative Feature Pruning with following parameters")
+        logger.log(20, f"max_num_fit: {max_num_fit}, prune_threshold: {prune_threshold}, " +
+                       f"stop_threshold: {stop_threshold}, num_shuffle_sets: {num_shuffle_sets}")
 
+        is_oof = X_val is None or y_val is None
         best_score = None
         best_index = None
         fitted_copies_info = []
         current_features = None
+        self_copy = None
         try:
             for index in range(max_num_fit):
                 # Fit a deepcopied model on current_features
-                curr_model_path = os.path.join(self.path, f'prune_{index}', '')
                 self_copy = self.convert_to_template()
+                self_copy.rename(f'{self.name}_P{index}')
                 self_copy.features = current_features
-                self_copy.path = curr_model_path
                 self_copy.fit(X=X, y=y, X_val=X_val, y_val=y_val, **kwargs)
                 self_copy.save()
 
-                score = self_copy.score(X=X_val, y=y_val)
-                fitted_copies_info.append((score, curr_model_path))
+                # Score new model and decide whether to continue refitting
+                if is_oof:
+                    score = self_copy.score_with_oof(y)
+                    X_fi, y_fi = X, y
+                else:
+                    score = self_copy.score(X=X_val, y=y_val)
+                    X_fi, y_fi = X_val, y_val
+                fitted_copies_info.append((round(score, 4), self_copy.path))
                 if best_score is None or score >= best_score:
-                    logger.log(20, f"\tFit {index+1}: Current score {score} is better than best score {best_score}.")
+                    if best_score is None:
+                        logger.log(20, f"\tFit {index+1}: Current score is {score}")
+                    else:
+                        logger.log(20, f"\tFit {index+1}: Current score {score} is not worse than best score {best_score}.")
                     best_score = score
                     best_index = index
-                if index - best_index == stop_threshold:
+                if index - best_index >= stop_threshold:
                     logger.log(20, f"\tEnding prune loop after {index+1} iterations. Best score: {best_score}")
                     break
 
                 # compute all feature importance and remove ones that don't meet the threshold
-                importance_df = self_copy.compute_feature_importance(X=X_val, y=y_val, num_shuffle_sets=num_shuffle_sets)
+                importance_df = self_copy.compute_feature_importance(X=X_fi, y=y_fi, num_shuffle_sets=num_shuffle_sets)
```

Comment:
bug: is_oof must be specified here if it is True

Reply:
otherwise you will overfit massively by treating training data as validation data (not using OOF scores)

Reply:
Good catch. I should've checked myself rather than implementing what I thought you meant